Bump all APIs to v1

esoergel opened the pull request:
## Product Description
Make all APIs available in duplicate at v1 URLs

## Technical Summary
https://dimagi.atlassian.net/browse/USH-4789
As discussed in [[CEP] New API Versioning Scheme](https://docs.google.com/document/d/1Q-uOTgbpRMSTxZ4o6aThcGb5weNyMiVyU_TtJyR3O50/edit)

## Feature Flag


## Safety Assurance


### Safety story


### Automated test coverage


### QA Plan


### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Put resource_name in URL path at the last minute

This will make it possible for us to support both formats:

  OLD: v0.5/resource_name
  NEW: resource_name/v1

esoergel committed:
Add system for new style API versioning

esoergel committed:
Ensure correct api_name is accessible

```
>>> from corehq.apps.api.resources.v0_5 import CommCareUserResource
>>> v1 = CommCareUserResource(api_name="v1")
>>> v5 = CommCareUserResource(api_name="v0.5")
>>> v5.api_name
'v0.5'
>>> v1.api_name
'v1'
>>> # That worked fine, but here's what tastypie uses:
>>> v5._meta.api_name
'v1'
>>> v1._meta.api_name
'v1'
>>> # and here's why:
>>> v1._meta is v5._meta
True
```
Tastypie supports passing in `api_name` to the resource already, but if
you don't, it gets set in the `API.register` method, also on the `_meta`
object, so it's the same issue in both instances

esoergel committed:
Copy all v0.5 resources to v1

esoergel committed:
Copy the couple new v0.6 resources to v2

esoergel committed:
These three don't extend HqBaseResource; use mixin

I suspect these probably _should_ extend the base resource, but I don't
want to grow the scope of my changes like that

esoergel committed:
Add basic tests for URL mapping

esoergel committed:
Pull out the odata urlpatterns

esoergel committed:
Convert to more traditional list of URLs

Also kinda thinking I should reduce the scope of the `get_urlpattern`
thing I added, pulling it in-line here, even if it means a bit of
unnecessary duplication, just for familiarity

esoergel committed:
Make non-tastypie URLs also available at v1

One hesitation I have about this change is that it modifies the
URL returned by `reverse` for each of these to the new latest version.
That means for example, that those of these which return a 'next'
parameter will have it now pointing to the newer version.  The view is
the same in both instances, so the only way I can imagine it breaking is
if the client has something that checks the format of the URL returned.

esoergel committed:
Simplify global API spec a tad

esoergel committed:
Bump user APIs to v1

esoergel committed:
docstring

esoergel committed:
lint

esoergel commented on corehq/apps/api/urls.py:
```diff
@@ -1,3 +1,28 @@
+
+    /a/mydomain/api/user/v1/
+
+To manage the transition, all v0.5 URls are duplicated as v1, and v0.6 URLs are
+duplicated as v2.
```
(copying here from one of my commit messages)

One hesitation I have about this change is that it modifies the URL returned by `reverse` for each of the non tastypie views to the new latest version. That means for example, that those of these which return a 'next' parameter will have it now pointing to the newer version.  The view is the same in both instances, so the only way I can imagine it breaking is if the client has something that checks the format of the URL returned, but I don't like the behavior change.

This isn't the case where there's actually a separate view for each version, only when both versions use the same exact code and url name.  Actually this isn't a problem with tastypie resources, as the version number is one of the URL kwargs that must be passed in to reverse, allowing the URL resolver to disambiguate.  I could probably adopt a similar pattern elsewhere if this is a concern, it just means changing each of the views individually.

esoergel commented on corehq/apps/api/urls.py:
```diff
@@ -94,47 +114,63 @@ def top_level(self, request, api_name=None, **kwargs):
+    url(r'v0\.6/case(?:/(?P<case_id>[\w\-,]+))?/?$', case_api),
+    url(r'case/v2(?:/(?P<case_id>[\w\-,]+))?/?$', case_api, name='case_api'),
+    path('', include(list(versioned_apis(_OLD_API_LIST)))),
+    url(r'^case/attachment/(?P<case_id>[\w\-:]+)/(?P<attachment_id>.*)$', CaseAttachmentAPI.as_view(),
+        name="api_case_attachment"),
+    url(r'^form/attachment/(?P<instance_id>[\w\-:]+)/(?P<attachment_id>.*)$', view_form_attachment,
+        name="api_form_attachment"),
```
Good question - I don't have a strong opinion.  I was picturing them as unversioned resources, since they just return blobs, but I suppose there's always room to grow.  What's your preference?  If we went with a v1 number, we'd need to decide which to go with:
`form/v1/attachment/instance_id/attachment_id`
`form/attachment/v1/instance_id/attachment_id`
`form_attachment/v1/instance_id/attachment_id`

The regular forms API is `form/v1`, so there's a semantic question of whether this is part of that API or an independent resource.

snopoke replied:
My preference would be to rename them (option 3 in your list). They aren't documented anywhere currently either.

esoergel committed:
version form and case attachments resources

This also makes it clearer that they are resources independent of form/
and case/, and versioned separately

esoergel committed:
update odata api URLs

esoergel committed:
Update user repeater to v1 url

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into es/api-v1

esoergel committed:
Ensure UCR API reverses to same URL pattern

I could also pass the arg through from the function all the way down,
but it's also available on the `request`, so I dunno

esoergel committed:
Make messaging-event API respect URL pattern

esoergel committed:
Use proper version when reversing odata URLs

esoergel committed:
Use consistent v1 URL pattern for odata

the "resource names" are `odata/cases` and `odata/forms`, so ensure that
that is consistent between tastypie and non-tastypie views

esoergel committed:
update test URL resolving

esoergel committed:
lint

esoergel commented:
Just pushed some commits here to address the APIs I found which contain URLs to themselves - I wanted to make sure accessing the v0.5 URL for example still returns the v0.5 URL in the `next` parameter and that sort of thing.

esoergel committed:
Update odata link to v1
